### PR TITLE
i#6662 public traces, part 3: regdeps disasm

### PR DESCRIPTION
**Pull request by edeiana**
Adds disassembling for DR_ISA_REGDEPS instructions.
Specifically, when we disassemble DR_ISA_REGDEPS instructions, we print the 
instruction encoding (divided in 4 byte words, can span one or two lines, similar to x86),
we substitute the opcode with categories, we print the operations size (e.g., `[4byte]`),
and then the source and destination virtual register names (e.g., `%rv3`).

Disassembled instructions look as follows:
`00000812 06260606 load [8byte]       %rv4 -> %rv4 %rv36`

In general, they follow this pattern:
[encoding (in 4 byte words)]  [categories]  [operation_size]  [src_regs -> dst_regs]

Issue: #6662

### CODE REVIEW

File: core/ir/disassemble_shared.c [hunk 1]
```diff
@@ -1164,6 +1224,48 @@ sign_extend_immed(instr_t *instr, int srcnum, opnd_t *src)
     }
 }
 
+#define MAX_LENGTH_ALL_CATEGORIES_AS_STRING 57
```
Comment (edeiana) [inline core/ir/disassemble_shared.c:1227]: Perhaps this and the following two functions should be moved to a new file: core/ir/isa_regdeps/disassemble.c.

**Comment by edeiana** [comment]
There are 2 tests consistently failing in ci-x86 / x86-64: code_api|tool.drcacheoff.skip and code_api|tool.drcacheoff.skip2.
The cause is that the header of the trace is being removed and that is causing the dcontext ISA mode to be set to DR_ISA_REGDEPS (on a, supposedly, x86 trace).
I need to investigate why it's happening.

File: core/ir/disassemble_shared.c [hunk 2]
```diff
@@ -749,9 +751,15 @@ internal_opnd_disassemble(char *buf, size_t bufsz, size_t *sofar DR_PARAM_INOUT,
...
         case PC_kind:
         case FAR_PC_kind: break;
-        case REG_kind:
-            if (!opnd_is_reg_partial(opnd))
+        case REG_kind: {
+            /* We always want to print the operation size for DR_ISA_REGDEPS
+             * instructions.
+             */
+            bool is_global_isa_mode_synthetic =
+                dr_get_isa_mode(get_thread_private_dcontext()) == DR_ISA_REGDEPS;
```
Comment (derekbruening) [inline core/ir/disassemble_shared.c:759]: Hmm, this overhead may well show up and slow down disassembly.  Don't we already have a dcontext passed in?  Use that instead of calling get_thread_*.  Even then, can we avoid this extra work on every single register by setting use_size_sfx differently in some way?  Hmm.

So for the operation size we have just one size for the largest source size, right?  Should we print that next to the categories instead of on every register?  It might be misleading on every register since that's not what the size means.

File: core/ir/disassemble_shared.c [hunk 3]
```diff
@@ -809,6 +817,48 @@ print_extra_bytes_to_file(file_t outfile, byte *pc, byte *next_pc, int extra_sz,
...
+        extra_sz = 0;
+    for (i = 0; i < sz; i++)
+        print_to_buffer(buf, bufsz, sofar, " %02x", *(pc + i));
+    for (i = sz; i < REGDEPS_BYTES_PER_LINE; i++)
+        print_to_buffer(buf, bufsz, sofar, "   ");
+    print_to_buffer(buf, bufsz, sofar, " ");
+    return extra_sz;
+}
+
+void
```
Comment (derekbruening) [inline core/ir/disassemble_shared.c:845]: static

File: core/ir/disassemble_shared.c [hunk 4]
```diff
@@ -809,6 +817,48 @@ print_extra_bytes_to_file(file_t outfile, byte *pc, byte *next_pc, int extra_sz,
     os_write(outfile, buf, sofar);
 }
 
+/* DR_ISA_REGDEPS instruction encodings can be at most 16 bytes,
+ * hence we can have at most 2 lines.
+ */
+#define REGDEPS_BYTES_PER_LINE 8
+
+int
```
Comment (derekbruening) [inline core/ir/disassemble_shared.c:825]: static (don't make global functions without a good reason)

File: core/ir/disassemble_shared.c [hunk 5]
```diff
@@ -809,6 +817,48 @@ print_extra_bytes_to_file(file_t outfile, byte *pc, byte *next_pc, int extra_sz,
...
+{
+    int sz = (int)(next_pc - pc);
+    int i, extra_sz;
+    if (sz > REGDEPS_BYTES_PER_LINE) {
+        extra_sz = sz - REGDEPS_BYTES_PER_LINE;
+        sz = REGDEPS_BYTES_PER_LINE;
+    } else
+        extra_sz = 0;
+    for (i = 0; i < sz; i++)
+        print_to_buffer(buf, bufsz, sofar, " %02x", *(pc + i));
```
Comment (derekbruening) [inline core/ir/disassemble_shared.c:838]: Do we want 1-byte separate bytes?  Every instruction has a 4-byte header: seems more natural to print a 4-byte word for that, right?

File: core/ir/disassemble_shared.c [hunk 6]
```diff
@@ -809,6 +817,48 @@ print_extra_bytes_to_file(file_t outfile, byte *pc, byte *next_pc, int extra_sz,
     os_write(outfile, buf, sofar);
 }
 
+/* DR_ISA_REGDEPS instruction encodings can be at most 16 bytes,
+ * hence we can have at most 2 lines.
+ */
+#define REGDEPS_BYTES_PER_LINE 8
+
+int
+print_regdeps_encoding_bytes_to_buffer(char *buf, size_t bufsz,
```
Comment (derekbruening) [inline core/ir/disassemble_shared.c:826]: Should these live in ir/isa_regdeps/ instead of here

File: core/ir/disassemble_shared.c (hunk 1 above)
Comment (derekbruening) [reply core/ir/disassemble_shared.c:1227]: +1, see comment above too.

File: core/ir/disassemble_shared.c [hunk 7]
```diff
@@ -1164,6 +1224,48 @@ sign_extend_immed(instr_t *instr, int srcnum, opnd_t *src)
...
+
+    /* We consider 0x80000000 enough to be future proof when adding new categories.
+     */
+    const uint max_mask = 0x80000000;
+    uint start_index = 0;
+    for (uint mask = 0x1; mask <= max_mask; mask <<= 1) {
+        if (TESTANY(mask, category)) {
+            const char *category_name = instr_get_category_name(mask);
+            size_t char_to_copy = strlen(category_name);
+            strncpy(category_names + start_index, category_name, char_to_copy);
```
Comment (derekbruening) [inline core/ir/disassemble_shared.c:1251]: This should be able to use print_to_buffer() which handles all the updating of appending to a buffer.

File: core/ir/disassemble_shared.c [hunk 8]
```diff
@@ -1164,6 +1224,48 @@ sign_extend_immed(instr_t *instr, int srcnum, opnd_t *src)
...
 
+#define MAX_LENGTH_ALL_CATEGORIES_AS_STRING 57
+
+/* Returns a space-separated string representation of the list of categories an
+ * instruction belongs to.
+ */
+void
+get_category_names(uint category, char category_names[], size_t category_names_size)
+{
+    memset(category_names, '\0', category_names_size);
```
Comment (derekbruening) [inline core/ir/disassemble_shared.c:1235]: This seems inefficient: we just need a 0 at the end of the copy.  Normally we use standard helpers like print_to_buffer() or snprintf() + NULL_TERMINATE_BUFFER.

File: core/ir/disassemble_shared.c [hunk 9]
```diff
@@ -1164,6 +1224,48 @@ sign_extend_immed(instr_t *instr, int srcnum, opnd_t *src)
...
+    if (category == DR_INSTR_CATEGORY_UNCATEGORIZED) {
+        const char *category_name = instr_get_category_name(category);
+        size_t char_to_copy = strlen(category_name);
+        strncpy(category_names, category_name, char_to_copy);
```
Comment (derekbruening) [inline core/ir/disassemble_shared.c:1239]: This is not checking category_names_size; nor is it following https://dynamorio.org/page_code_content.html#autotoc_md109 which says "Always null-terminate buffers after calling `strncpy` or `snprintf`: for statically-sized buffers, use  NULL_TERMINATE_BUFFER()."

File: core/ir/disassemble_shared.c (hunk 2 above)
Comment (derekbruening) [inline core/ir/disassemble_shared.c:759]: I thought the mode of a decoded instruction was stored in the instr_t and not in the dcontext.  Maybe the instr_t needs to be passed in here.

File: clients/drcachesim/tools/view.cpp [hunk 10]
```diff
@@ -50,6 +50,7 @@
 
 #include "analysis_tool.h"
 #include "dr_api.h"
+#include "dr_ir_encode.h"
```
Comment (derekbruening) [inline clients/drcachesim/tools/view.cpp:53]: Mixed feelings about clang-tidy's checks for headers which seem to go too far and do not allow how the DR headers were designed with dr_api.h including the others (this is also how some system headers work, along with other API headers most notably windows.h).  IMHO this include should not be necessary as it goes against the recommended usage in the DR docs.  We may be able to find a way to configure clang-tidy.

File: clients/drcachesim/tests/view_test.cpp [hunk 11]
```diff
@@ -586,11 +586,63 @@ run_chunk_tests(void *drcontext)
...
+        std::string(R"DELIM(           1           0:           3 <marker: version 3>
+           2           0:           3 <marker: filetype 0x20e00>
+           3           0:           3 <marker: cache line size 64>
+           4           0:           3 <marker: chunk instruction count 2>
+           5           0:           3 <marker: timestamp 1002>
+           6           0:           3 <marker: tid 3 on core 2>
+           7           1:           3 ifetch       )DELIM") +
+        std::string(R"DELIM(3 byte(s) @ 0x00007f6fdd3ec360 )DELIM") +
+        std::string(
+            R"DELIM(11 00 01 00 06 09 06 00 move   %rv4[8byte] -> %rv7[8byte]
```
Comment (derekbruening) [inline clients/drcachesim/tests/view_test.cpp:627]: Could we have more examples: what do loads/stores with memops look like; something with differing real sizes like SIMD with k mask and GPR in there too, which should really highlight whether sizes on registers are going to work (I suspect we can easily see a GPR with an impossible SIMD-sized size -- hence the suggestion to move these sizes to a single size after the categories or sthg like that.)

Comment (derekbruening) [inline clients/drcachesim/tests/view_test.cpp:627]: Probably there should be core IR tests independent of this drmemtrace view tool (could be a separate PR for the core IR, split from the drmemtrace view tool) in suite/test/api/.

File: core/ir/disassemble_shared.c (hunk 1 above)
Comment (edeiana) [reply core/ir/disassemble_shared.c]: Done.

File: core/ir/disassemble_shared.c (hunk 2 above)
Comment (edeiana) [reply core/ir/disassemble_shared.c]: This check has been removed, I set `use_size_sfx = 0` for regdeps instructions and print only one size (the operation) after the categories if it's not 0 (e.g., `branch` does not have an operation size).

File: core/ir/disassemble_shared.c (hunk 4 above)
Comment (edeiana) [reply core/ir/disassemble_shared.c]: Moved to core/ir/isa_regdeps/disassemble.c.

File: core/ir/disassemble_shared.c (hunk 5 above)
Comment (edeiana) [reply core/ir/disassemble_shared.c]: Yes, 4 byte words makes more sense. Done.

File: core/ir/disassemble_shared.c (hunk 6 above)
Comment (edeiana) [reply core/ir/disassemble_shared.c]: Moved to core/ir/isa_regdeps/disassemble.c.

File: core/ir/disassemble_shared.c (hunk 7 above)
Comment (edeiana) [reply core/ir/disassemble_shared.c]: Using `print_to_buffer()`. Done.

File: core/ir/disassemble_shared.c (hunk 8 above)
Comment (edeiana) [reply core/ir/disassemble_shared.c]: Using `print_to_buffer()`, which takes care of that. Done.

File: core/ir/disassemble_shared.c (hunk 9 above)
Comment (edeiana) [reply core/ir/disassemble_shared.c]: Using `print_to_buffer()`, which takes care of that. Done.

File: core/ir/disassemble_shared.c (hunk 2 above)
Comment (edeiana) [reply core/ir/disassemble_shared.c]: Removed the need for this check altogether.

File: clients/drcachesim/tools/view.cpp (hunk 10 above)
Comment (edeiana) [reply clients/drcachesim/tools/view.cpp]: Yes, clang-tidy is a bit strict. Removed.

File: clients/drcachesim/tests/view_test.cpp (hunk 11 above)
Comment (edeiana) [reply clients/drcachesim/tests/view_test.cpp]: Added load with from vector register %ymm to a GPR and a SIMD instruction with k mask and GPR.
Operation size is not a single value after the category names, instead of being printed for all register operands.

Comment (edeiana) [reply clients/drcachesim/tests/view_test.cpp]: Ok, will add tests in suite/test/api/dis* in a separate PR.

Comment (derekbruening) [reply clients/drcachesim/tests/view_test.cpp]: What I meant was one PR for core disassembly + tests, and a separate PR for the logically separate drmemtrace view tool then using the new disasm.  Changes to the core should have their own tests.  The drmemtrace tool is a logically separate piece.

File: clients/drcachesim/tests/view_test.cpp [hunk 12]
```diff
@@ -607,25 +611,49 @@ run_regdeps_test(void *drcontext)
...
+           4           0:           3 <marker: chunk instruction count 3>
            5           0:           3 <marker: timestamp 1002>
            6           0:           3 <marker: tid 3 on core 2>
            7           1:           3 ifetch       )DELIM") +
         std::string(R"DELIM(3 byte(s) @ 0x00007f6fdd3ec360 )DELIM") +
+        std::string(R"DELIM(00010011 move [8byte]       %rv4 -> %rv7
+)DELIM") +
+        std::string(R"DELIM(           7           1:           3)DELIM") +
+        std::string(
+            R"DELIM(                                             00060906
```
Comment (derekbruening) [inline clients/drcachesim/tests/view_test.cpp:638]: I'm finding this quite difficult to read when it split over all these lines with the DELIM in there.  It is hard to figure out what this "006*" is supposed to be, sitting here by itself with tons of spaces all around.  I think it would be best to have one matching line per output line even if it's long.

File: clients/drcachesim/tools/view.cpp [hunk 13]
```diff
@@ -573,14 +572,16 @@ view_t::parallel_shard_memref(void *shard_data, const memref_t &memref)
          * disassemble DR_ISA_REGDEPS instructions.
          */
         dr_isa_mode_t old_isa_mode = dr_get_isa_mode(dcontext_.dcontext);
-        if (TESTANY(OFFLINE_FILE_TYPE_ARCH_REGDEPS, filetype_)) {
+        bool is_regdeps_trace =
+            filetype_ != -1 && TESTANY(OFFLINE_FILE_TYPE_ARCH_REGDEPS, filetype_);
```
Comment (derekbruening) [inline clients/drcachesim/tools/view.cpp:576]: Change the init value to 0?  Not ideal to have to do this extra check.

File: clients/drcachesim/tools/view.cpp [hunk 14]
```diff
@@ -604,13 +605,16 @@ view_t::parallel_shard_memref(void *shard_data, const memref_t &memref)
         disasm.insert(newline, str.str());
     }
     // Put our prefix on raw byte spillover, and skip the other columns.
-    newline = disasm.find('\n');
-    if (newline != std::string::npos && newline < disasm.size() - 1) {
+    size_t pos = 0;
+    newline = disasm.find('\n', pos);
+    while (newline != std::string::npos && newline < disasm.size() - 1) {
```
Comment (derekbruening) [inline clients/drcachesim/tools/view.cpp:610]: This change goes away if it all fits on one or 2 lines?

File: core/ir/isa_regdeps/disassemble.h [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+ * DAMAGE.
+ */
+
+#ifndef _REGDEPS_DISASSEMBLE_H_
+#define _REGDEPS_DISASSEMBLE_H_ 1
+
+#include "../globals.h"
+
+int
+print_regdeps_encoding_bytes_to_buffer(char *buf, size_t bufsz,
```
Comment (derekbruening) [inline core/ir/isa_regdeps/disassemble.h:39]: In C with no namespaces, esp where we statically link to the app in some use cases and are prevented from hiding our global symbols, it's good to have a common prefix: so how about starting these global names with the new convention d_r_ or maybe d_r_regdeps_ or disassemble_ or d_r_disassemble_.

File: core/ir/isa_regdeps/disassemble.c [hunk 16]
```diff
@@ -0,0 +1,77 @@
...
+#include "../globals.h"
+#include "disassemble.h"
+
+/* DR_ISA_REGDEPS instruction encodings can be at most 16 bytes and are 4 byte aligned,
+ * hence we can have at most 4 lines of 4 bytes each.
+ */
+#define REGDEPS_BYTES_PER_LINE 4
+
+int
+print_regdeps_encoding_bytes_to_buffer(char *buf, size_t bufsz,
```
Comment (derekbruening) [inline core/ir/isa_regdeps/disassemble.c:42]: This function does not do what its name implies: it doesn't print all the bytes!  It just print 4 bytes!  Maybe best to rename ("d_r_regdeps_print_encoding_first_line"??), and probably add a comment explaining what it's doing and what it's returning, as that is not obvious.

Update: interesting, I see this parallels print_bytes_to_buffer: that also seems confusing.

File: core/ir/isa_regdeps/disassemble.c [hunk 17]
```diff
@@ -0,0 +1,77 @@
...
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+#include "../globals.h"
+#include "disassemble.h"
+
+/* DR_ISA_REGDEPS instruction encodings can be at most 16 bytes and are 4 byte aligned,
+ * hence we can have at most 4 lines of 4 bytes each.
```
Comment (derekbruening) [inline core/ir/isa_regdeps/disassemble.c:37]: Why are we limited to 4 bytes per line?  It doesn't seem great to have a 16-byte encoding with the first line with 4 encoding bytes and the actual disasm and then 3 following lines each with just 4 bytes of more encoding.  Seems better to either put the whole 16 bytes one one line or if that is too much empty space and 90% of the instructions are at <=8 bytes then put 8 bytes on each of 2 lines?

File: clients/drcachesim/tests/view_test.cpp (hunk 11 above)
Comment (edeiana) [reply clients/drcachesim/tests/view_test.cpp]: Oh, sorry about that, I missed the keyword: "split" in your first comment.
I'll split this PR in "disassembly + tests in suite/tests/api" and "view tool + current view_test.cpp on a regdeps trace".

File: clients/drcachesim/tests/view_test.cpp (hunk 12 above)
Comment (edeiana) [reply clients/drcachesim/tests/view_test.cpp:638]: Initially I did that, but DR wouldn't build because of the 90 char per line limit.
I thought it was a clang-format issue, so I tried:
```
/* clang-format off */
{ code with lines over 90 chars}
/* clang-format on */
```

But it didn't work.
I'll investigate what I need to turn off to have one matching line per output.

File: clients/drcachesim/tools/view.cpp (hunk 14 above)
Comment (edeiana) [reply clients/drcachesim/tools/view.cpp:610]: Yes, it would.

File: core/ir/isa_regdeps/disassemble.c (hunk 17 above)
Comment (edeiana) [reply core/ir/isa_regdeps/disassemble.c:37]: I thought 4 bytes per line was more similar, in terms of layout, to our description and examples of the encoding in
https://github.com/DynamoRIO/dynamorio/blob/a3411cb0eff4c6b6c1bff2dee37906baf8e7b457/core/ir/isa_regdeps/encoding_common.h#L84
since everything is in 4 byte chunks.

But we are not limited by it, I'm fine with having 8 bytes per line (maybe divided into two 4 byte words?) so we can stay within 2 lines instead of 4.

File: core/ir/isa_regdeps/disassemble.h (hunk 15 above)
Comment (edeiana) [reply core/ir/isa_regdeps/disassemble.h:39]: Ah, my bad, I forgot about the new convention we follow.
I'll go with: "d_r_disassemble_regdeps_".

File: clients/drcachesim/tools/view.cpp (hunk 13 above)
Comment (edeiana) [reply clients/drcachesim/tools/view.cpp:576]: But 0 is `OFFLINE_FILE_TYPE_DEFAULT`.
I'll check if it's safe to do so, I haven't considered the side-effects of that.

File: core/ir/isa_regdeps/disassemble.c (hunk 16 above)
Comment (edeiana) [reply core/ir/isa_regdeps/disassemble.c:42]: Right, I followed `print_bytes_to_buffer()`.
Perhaps I should rename both.

File: clients/drcachesim/tests/view_test.cpp (hunk 12 above)
Comment (derekbruening) [reply clients/drcachesim/tests/view_test.cpp:638]: That `clang-format off` is how to disable clang-format.  There are also vera++ checks but I forget what they cover these days (most added before we had clang-format in place): I do see make/style_checks/scripts/rules/line_length_content.tcl and line_length_content is enabled.  Add this file to the exclusion list in make/style_checks/scripts/rules/line_length_content.tcl?

Comment (edeiana) [reply clients/drcachesim/tests/view_test.cpp:638]: Ah, yes, that should do the trick, thank you Derek!

File: clients/drcachesim/tests/view_test.cpp (hunk 11 above)
Comment (edeiana) [reply clients/drcachesim/tests/view_test.cpp]: Done. This PR is only for core disassembly of regdeps instructions + tests.
The PR for the view tool + test is ready, once this one goes into master.

File: clients/drcachesim/tests/view_test.cpp (hunk 12 above)
Comment (edeiana) [reply clients/drcachesim/tests/view_test.cpp]: Not relevant to this PR anymore.

File: clients/drcachesim/tools/view.cpp (hunk 14 above)
Comment (edeiana) [reply clients/drcachesim/tools/view.cpp]: Not relevant to this PR anymore. Fixed in view tool PR.

File: core/ir/isa_regdeps/disassemble.h (hunk 15 above)
Comment (edeiana) [reply core/ir/isa_regdeps/disassemble.h]: Done.
I went with you suggestion in the end: `d_r_regdeps_print_encoding_first_line()` (mine was getting too long).

File: core/ir/isa_regdeps/disassemble.c (hunk 16 above)
Comment (edeiana) [reply core/ir/isa_regdeps/disassemble.c]: Renaming done, also for the `print_[extra_]bytes_to_buffer()` routines.
Now: `d_r_print_encoding_first_line_to_buffer()` and `d_r_print_encoding_second_line_to_buffer()`.

File: core/ir/isa_regdeps/disassemble.c (hunk 17 above)
Comment (edeiana) [reply core/ir/isa_regdeps/disassemble.c]: Done. Went with 2 lines max, 8 bytes each, split in two 4 byte words (if the second/fourth word is present).

File: suite/tests/api/ir_regdeps.c [hunk 18]
```diff
@@ -209,13 +250,16 @@ test_instr_create_encode_decode_synthetic_x86_64(void *dc)
...
     decode(dc, buf, instr);
-    test_instr_encode_decode_synthetic(dc, instr);
+    const char *expected_disasm_str_vdpbf16ps_mask =
+        " 00000031 65646640 uncategorized [64byte]       %rv98 %rv99 %rv102 -> %rv100\n "
+        "00000068";
+    test_instr_encode_decode_synthetic(dc, instr, expected_disasm_str_vdpbf16ps_mask);
 }
 #endif
 
 #ifdef ARM
```
Comment (edeiana) [inline suite/tests/api/ir_regdeps.c:260]: Both ARM and RISCV64 tests build but are not being run (looking at Github tests: "ci-aarchxx-cross / arm-cross-compile" and "ci-riscv64 / riscv64-cross-compile" respectively).
Are my define wrong? (But I believe those are what's used in other tests as well)
Or am I doing something wrong in CMakeLists.txt that I'm not seeing?

https://github.com/DynamoRIO/dynamorio/blob/a559edc0b4c6a2ac4c2c94702a6a7bf87e408533/suite/tests/CMakeLists.txt#L2040

File: core/ir/isa_regdeps/disassemble.c (hunk 16 above)
Comment (derekbruening) [reply core/ir/isa_regdeps/disassemble.c]: SGTM

File: core/ir/isa_regdeps/disassemble.c (hunk 17 above)
Comment (derekbruening) [reply core/ir/isa_regdeps/disassemble.c]: SGTM

File: suite/tests/api/ir_regdeps.c (hunk 18 above)
Comment (derekbruening) [reply suite/tests/api/ir_regdeps.c:260]: The cross runs only execute a short list of tests known to succeed under QEMU.  Look for the RUNS_ON_QEMU test lists at the bottom of the file.

File: suite/tests/api/ir_regdeps.c [hunk 19]
```diff
@@ -136,62 +160,79 @@ test_instr_create_encode_decode_synthetic_x86_64(void *dc)
     instr_encode(dc, instr, buf);
     instr_reset(dc, instr);
     decode(dc, buf, instr);
-    test_instr_encode_decode_synthetic(dc, instr);
+    const char *expected_disasm_str_push =
+        " 00001021 26060606 store [8byte]       %rv4 %rv36 -> %rv4\n";
```
Comment (derekbruening) [inline suite/tests/api/ir_regdeps.c:164]: (Style looks good to me.)

**Review by derekbruening — APPROVED** [review]

File: suite/tests/api/ir_regdeps.c (hunk 18 above)
Comment (edeiana) [reply suite/tests/api/ir_regdeps.c:260]: Ah, that's what I was missing, thanks.

File: clients/drcachesim/tools/view.cpp (hunk 13 above)
Comment (edeiana) [reply clients/drcachesim/tools/view.cpp]: Not relevant to this PR anymore.

File: suite/tests/api/ir_regdeps.c (hunk 18 above)
Comment (edeiana) [reply suite/tests/api/ir_regdeps.c:267]: I enabled the `api.ir_regdeps` test for APPLE (still runs on x86_64, so no changes needed) and RISCV64 (categories there are all "uncategorized", but better than nothing).
Couldn't enable it for ARM, as I get: "/lib/ld-linux-armhf.so.3: No such file or directory", so that's disabled (added a FIXME comment about it, if we want to enable it in the future).
